torch/distributed/checkpoint/state_dict.py

```diff
@@ -555,19 +555,34 @@ def _load_model_state_dict(
     assign = False
     if info.broadcast_from_rank0 or info.full_state_dict:
         device = None
+        devices = set()
         for key, value in local_state_dict.items():
             if torch.is_tensor(value) and value.dim() > 0:
-                if device is None:
-                    device = value.device
-                else:
-                    assert device == value.device
-        assert device is not None
-        if device == torch.device("meta"):
-            device = dist.distributed_c10d._get_pg_default_device()
-            assign = True
+                devices.add(value.device)
+        if len(devices) == 1:
+            device = devices.pop()
+            if device == torch.device("meta"):
+                device = dist.distributed_c10d._get_pg_default_device()
+                assign = True
```

Comment:
Do you think that if we always do `assign=True`, is it better?

Reply:
test_shared_weight would fail id we set assign to True by default.
We can change assign and it's related tests in the future PR.